### PR TITLE
Fix prefix store seg fault

### PR DESCRIPTION
**Pull request by H-Huang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #136872

fixes https://github.com/pytorch/pytorch/issues/136723

Do not allow `None` to be passed into `PrefixStore`

cc @XilunWu @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

### CODE REVIEW

File: test/distributed/test_store.py [hunk 1]
```diff
@@ -250,6 +250,8 @@ def test_get_underlying_store(self):
                 prefix_store = dist.PrefixStore("prefix", store)
                 self.assertEqual(prefix_store.underlying_store, store)
 
+        with self.assertRaises(ValueError):
+            dist.PrefixStore("prefix", None)
```
Comment (kwen2501) [inline test/distributed/test_store.py:254]: nit: add comment?

File: torch/csrc/distributed/c10d/init.cpp [hunk 2]
```diff
@@ -1588,7 +1588,16 @@ that adds a prefix to each key inserted to the store.
     prefix (str): The prefix string that is prepended to each key before being inserted into the store.
     store (torch.distributed.store): A store object that forms the underlying key-value store.
       )")
-      .def(py::init<const std::string&, c10::intrusive_ptr<::c10d::Store>>())
```
Comment (kwen2501) [inline torch/csrc/distributed/c10d/init.cpp:1591]: The previous behavior looks very suspicious that makes me even wonder if there is a specific purpose. (lol)

**Review by kwen2501 — APPROVED** [review]
Looks good to me. Minor comments. 
And de-stack from previous PR if not too much a pain.

File: torch/csrc/distributed/c10d/init.cpp (hunk 2 above)
Comment (H-Huang) [reply torch/csrc/distributed/c10d/init.cpp:1591]: im not sure either hahaha, lets see from the tests. Though there were very few PrefixStore tests from what i can tell

**Pull request closed without merging**